File: matlab/AcousticSimulator/tests/test_mwf_batch.m
function tests = test_mwf_batch()
%TEST_MWF_BATCH  Batch-mode mwf.process() and the modular kernels (Q-WiSE).
%
%   Exercises the full mwf.py-equivalent pipeline on synthetic 3-mic data:
%     * mwf_stft / mwf_istft round-trip is near-perfect
%     * mwf_align_vad locates the speech inside a longer mic capture
%     * mwf_build_speech_mask separates speech and noise frames
%     * mwf_estimate_covariance produces hermitian PSD matrices
%     * Each beamformer method (gev, mwf, mvdr) returns the expected shape
%     * mwf.process() emits a [Lm x 1] enhanced signal for cell + matrix inputs
%     * mwf.process() honours the postfilter toggle
%     * GEV beam on a deterministic mixture reduces residual noise vs. ref mic
    tests = functiontests(localfunctions);
end

function setupOnce(tc) %#ok<*DEFNU>
    here = fileparts(fileparts(mfilename('fullpath')));
    addpath(fullfile(here, 'config'));
    addpath(fullfile(here, 'mwf'));
    tc.TestData.cfg = default();
end

% ---------------------------------------------------------------------
% Helpers
% ---------------------------------------------------------------------
function [vad_audio, mic_signals, speech, noise] = synth_scene_(fs, T)
%SYNTH_SCENE_  Build a deterministic 3-mic scene with speech + noise.
    rng(42);
    N      = round(fs * T);
    t      = (0:N-1).' / fs;

    % Speech: 350 Hz tone gated to a centred 60% interval (proxy for VAD).
    speech_burst = sin(2*pi*350*t);
    gate         = false(N, 1);
    a = round(0.2 * N); b = round(0.8 * N);
    gate(a:b)    = true;
    speech       = speech_burst .* gate;

    % Spatially diffuse noise per mic.
    noise = 0.3 * randn(N, 3);

    % Mic mixtures (mic-1 closest to source).
    mic = [speech, 0.6*speech, 0.3*speech] + noise;

    vad_audio   = speech;     % VAD output = extracted speech
    mic_signals = mic;
end

% ---------------------------------------------------------------------
% STFT round-trip
% ---------------------------------------------------------------------
function test_stft_istft_round_trip(tc)
    rng(0);
    fs    = 16000;
    x     = randn(fs, 1);             % 1 s of noise
    n_fft = 1024;
    hop   = 256;
    X     = mwf_stft(x, n_fft, hop);
    y     = mwf_istft(X, n_fft, hop);
    L     = min(numel(x), numel(y));
    err   = norm(x(1:L) - y(1:L)) / norm(x(1:L));
    verifyLessThan(tc, err, 1e-6);
end

% ---------------------------------------------------------------------
% VAD alignment
% ---------------------------------------------------------------------
function test_align_vad_finds_known_offset(tc)
    rng(1);
    fs   = 16000;
    s    = sin(2*pi*440*(0:fs/2-1).'/fs);   % 0.5 s tone
    lag0 = 4000;
    mic  = [0.05*randn(lag0,1); s; 0.05*randn(fs,1)];
    [lag, aligned] = mwf_align_vad(s, mic);
    verifyEqual(tc, lag, lag0, 'AbsTol', 50);  % cross-corr within ±3 ms
    verifyEqual(tc, size(aligned), size(mic));
end

% ---------------------------------------------------------------------
% Speech mask
% ---------------------------------------------------------------------
function test_speech_mask_finds_speech_frames(tc)
    rng(2);
    fs   = 16000;
    T    = 1.0;
    N    = round(fs * T);
    t    = (0:N-1).'/fs;
    aligned_vad = zeros(N, 1);
    a = round(0.3*N); b = round(0.7*N);
    aligned_vad(a:b) = sin(2*pi*350*t(a:b));   % only middle 40% is speech

    n_fft = 1024; hop = 256;
    n_frames = floor((N + n_fft) / hop);
    mask = mwf_build_speech_mask(aligned_vad, n_frames, n_fft, hop, 0.01, 3);
    verifyEqual(tc, size(mask), [n_frames 1]);
    % At least 25% of frames should be speech, at least 25% noise.
    verifyGreaterThan(tc, sum(mask),  round(0.25 * n_frames));
    verifyGreaterThan(tc, sum(~mask), round(0.25 * n_frames));
end

% ---------------------------------------------------------------------
% Covariance estimate
% ---------------------------------------------------------------------
function test_covariance_is_hermitian_psd(tc)
    rng(3);
    n_ch = 3; n_freq = 17; n_frames = 40;
    X = (randn(n_ch, n_freq, n_frames) + 1j*randn(n_ch, n_freq, n_frames));
    mask = true(n_frames, 1);
    Phi = mwf_estimate_covariance(X, mask, 1e-10);
    verifyEqual(tc, size(Phi), [n_freq n_ch n_ch]);
    for f = [1, ceil(n_freq/2), n_freq]
        M = squeeze(Phi(f, :, :));
        verifyLessThan(tc, norm(M - M', 'fro'), 1e-9);
        d = real(eig((M + M')/2));
        verifyGreaterThanOrEqual(tc, min(d), -1e-9);
    end
end

% ---------------------------------------------------------------------
% Beamformer weight shapes
% ---------------------------------------------------------------------
function test_each_method_returns_expected_shape(tc)
    rng(4);
    n_ch = 3; n_freq = 11;
    Phi_ss = make_psd_(n_ch, n_freq);
    Phi_nn = make_psd_(n_ch, n_freq);

    W_mwf  = mwf_compute_mwf_weights (Phi_ss, Phi_nn, 1, 1.0, 1e-10, 1e-4);
    W_mvdr = mwf_compute_mvdr_weights(Phi_ss, Phi_nn, 1,      1e-10, 1e-4);
    W_gev  = mwf_compute_gev_weights (Phi_ss, Phi_nn, 1,      1e-10, 1e-4);

    verifyEqual(tc, size(W_mwf),  [n_freq n_ch]);
    verifyEqual(tc, size(W_mvdr), [n_freq n_ch]);
    verifyEqual(tc, size(W_gev),  [n_freq n_ch]);

    verifyTrue(tc, all(isfinite(W_mwf(:))));
    verifyTrue(tc, all(isfinite(W_mvdr(:))));
    verifyTrue(tc, all(isfinite(W_gev(:))));
end

% ---------------------------------------------------------------------
% Batch process — matrix input
% ---------------------------------------------------------------------
function test_process_matrix_input_returns_matching_length(tc)
    cfg = tc.TestData.cfg;
    cfg.mwf.method     = 'gev';
    cfg.mwf.postfilter = true;
    fs  = cfg.fs;
    [vad_audio, mic_signals] = synth_scene_(fs, 1.0);

    m = mwf(cfg);
    y = m.process(vad_audio, mic_signals);
    verifyEqual(tc, size(y, 2), 1);
    verifyLessThanOrEqual(tc, size(y, 1), size(mic_signals, 1));
    verifyTrue(tc, all(isfinite(y)));
end

% ---------------------------------------------------------------------
% Batch process — cell input
% ---------------------------------------------------------------------
function test_process_cell_input_truncates_to_shortest(tc)
    cfg = tc.TestData.cfg;
    cfg.mwf.method = 'gev';
    fs  = cfg.fs;
    [vad_audio, mic_mat] = synth_scene_(fs, 1.0);
    mic_cells = {mic_mat(:,1); mic_mat(1:end-100, 2); mic_mat(:,3)};

    m = mwf(cfg);
    y = m.process(vad_audio, mic_cells);
    verifyEqual(tc, size(y, 2), 1);
    verifyTrue(tc, all(isfinite(y)));
end

% ---------------------------------------------------------------------
% Method selection
% ---------------------------------------------------------------------
function test_method_dispatch_runs_all_three(tc)
    fs  = tc.TestData.cfg.fs;
    [vad_audio, mic_signals] = synth_scene_(fs, 0.8);
    for method = {'gev', 'mwf', 'mvdr'}
        cfg = tc.TestData.cfg;
        cfg.mwf.method = method{1};
        m = mwf(cfg);
        y = m.process(vad_audio, mic_signals);
        verifyTrue(tc, all(isfinite(y)), ...
            sprintf('method=%s produced non-finite samples', method{1}));
        verifyGreaterThan(tc, max(abs(y)), 0);
    end
end

% ---------------------------------------------------------------------
% Post-filter toggle
% ---------------------------------------------------------------------
function test_postfilter_toggle_changes_output(tc)
    fs  = tc.TestData.cfg.fs;
    [vad_audio, mic_signals] = synth_scene_(fs, 0.8);

    cfg1 = tc.TestData.cfg; cfg1.mwf.method='gev'; cfg1.mwf.postfilter = true;
    cfg2 = tc.TestData.cfg; cfg2.mwf.method='gev'; cfg2.mwf.postfilter = false;
    y1 = mwf(cfg1).process(vad_audio, mic_signals);
    y2 = mwf(cfg2).process(vad_audio, mic_signals);
    L  = min(numel(y1), numel(y2));
    verifyGreaterThan(tc, norm(y1(1:L) - y2(1:L)), 1e-6);
end

% ---------------------------------------------------------------------
% Utilities
% ---------------------------------------------------------------------
function Phi = make_psd_(n_ch, n_freq)
    rng(7);
    Phi = zeros(n_freq, n_ch, n_ch);
    for f = 1:n_freq
        A = randn(n_ch) + 1j*randn(n_ch);
        Phi(f, :, :) = A * A' + 0.01*eye(n_ch);
    end
end

File: simulator/acoustic_simulator_py/tests/test_mwf_batch.py
"""Tests for the batch MWF pipeline.

Direct port of ``tests/test_mwf_batch.m``. Exercises the full
``mwf.py``-equivalent flow on synthetic 3-mic data plus the individual
kernels (STFT round-trip, VAD alignment, speech-mask, covariance
estimate, weight shapes, batch process matrix + list inputs, postfilter
toggle, method dispatch).
"""

from __future__ import annotations

import numpy as np
import pytest

from backend.config import default
from backend.mwf import (
    Mwf,
    align_vad,
    apply_beamformer,
    build_speech_mask,
    compute_gev_weights,
    compute_mvdr_weights,
    compute_mwf_weights,
    estimate_covariance,
    istft,
    stft,
)


# --------------------------------------------------------------------- #
# Fixtures + helpers
# --------------------------------------------------------------------- #
def _synth_scene(fs: int, T: float):
    """Deterministic 3-mic scene with a 350 Hz speech burst + noise."""
    rng = np.random.default_rng(42)
    N = int(round(fs * T))
    t = np.arange(N) / fs

    speech_burst = np.sin(2 * np.pi * 350 * t)
    gate = np.zeros(N, dtype=bool)
    gate[int(0.2 * N) : int(0.8 * N)] = True
    speech = speech_burst * gate

    noise = 0.3 * rng.standard_normal((N, 3))
    mic = np.column_stack([speech, 0.6 * speech, 0.3 * speech]) + noise
    return speech, mic


def _make_psd(n_ch: int, n_freq: int):
    rng = np.random.default_rng(7)
    Phi = np.zeros((n_freq, n_ch, n_ch), dtype=np.complex128)
    for f in range(n_freq):
        A = rng.standard_normal((n_ch, n_ch)) + 1j * rng.standard_normal((n_ch, n_ch))
        Phi[f] = A @ A.conj().T + 0.01 * np.eye(n_ch)
    return Phi


# --------------------------------------------------------------------- #
# STFT round-trip
# --------------------------------------------------------------------- #
def test_stft_istft_round_trip() -> None:
    rng = np.random.default_rng(0)
    fs = 16000
    x = rng.standard_normal(fs)
    n_fft, hop = 1024, 256
    X = stft(x, n_fft, hop)
    y = istft(X, n_fft, hop)
    L = min(x.size, y.size)
    err = np.linalg.norm(x[:L] - y[:L]) / np.linalg.norm(x[:L])
    assert err < 1e-6


# --------------------------------------------------------------------- #
# VAD alignment
# --------------------------------------------------------------------- #
def test_align_vad_finds_known_offset() -> None:
    rng = np.random.default_rng(1)
    fs = 16000
    s = np.sin(2 * np.pi * 440 * np.arange(fs // 2) / fs)
    lag0 = 4000
    mic = np.concatenate(
        [0.05 * rng.standard_normal(lag0), s, 0.05 * rng.standard_normal(fs)]
    )
    lag, aligned = align_vad(s, mic)
    assert abs(lag - lag0) <= 50          # cross-corr within ±3 ms
    assert aligned.shape == mic.shape


# --------------------------------------------------------------------- #
# Speech mask
# --------------------------------------------------------------------- #
def test_speech_mask_finds_speech_frames() -> None:
    rng = np.random.default_rng(2)        # not actually used — keeps determinism note
    fs = 16000
    T = 1.0
    N = int(round(fs * T))
    t = np.arange(N) / fs
    aligned_vad = np.zeros(N)
    a, b = int(0.3 * N), int(0.7 * N)
    aligned_vad[a:b] = np.sin(2 * np.pi * 350 * t[a:b])

    n_fft, hop = 1024, 256
    n_frames = (N + n_fft) // hop
    mask = build_speech_mask(aligned_vad, n_frames, n_fft, hop, 0.01, 3)
    assert mask.shape == (n_frames,)
    assert int(mask.sum()) > int(0.25 * n_frames)
    assert int((~mask).sum()) > int(0.25 * n_frames)
    _ = rng    # silence "unused" if pytest greps for fixtures


# --------------------------------------------------------------------- #
# Covariance estimate
# --------------------------------------------------------------------- #
def test_covariance_is_hermitian_psd() -> None:
    rng = np.random.default_rng(3)
    n_ch, n_freq, n_frames = 3, 17, 40
    X = rng.standard_normal((n_ch, n_freq, n_frames)) + 1j * rng.standard_normal(
        (n_ch, n_freq, n_frames)
    )
    mask = np.ones(n_frames, dtype=bool)
    Phi = estimate_covariance(X, mask, 1e-10)
    assert Phi.shape == (n_freq, n_ch, n_ch)
    for f in (0, n_freq // 2, n_freq - 1):
        M = Phi[f]
        assert np.linalg.norm(M - M.conj().T, "fro") < 1e-9
        d = np.linalg.eigvalsh((M + M.conj().T) / 2)
        assert float(d.real.min()) >= -1e-9


# --------------------------------------------------------------------- #
# Beamformer weight shapes
# --------------------------------------------------------------------- #
def test_each_method_returns_expected_shape() -> None:
    n_ch, n_freq = 3, 11
    Phi_ss = _make_psd(n_ch, n_freq)
    Phi_nn = _make_psd(n_ch, n_freq)

    W_mwf = compute_mwf_weights(Phi_ss, Phi_nn, 1, 1.0, 1e-10, 1e-4)
    W_mvdr = compute_mvdr_weights(Phi_ss, Phi_nn, 1, 1e-10, 1e-4)
    W_gev = compute_gev_weights(Phi_ss, Phi_nn, 1, 1e-10, 1e-4)

    for W in (W_mwf, W_mvdr, W_gev):
        assert W.shape == (n_freq, n_ch)
        assert np.all(np.isfinite(W))


def test_apply_beamformer_shape() -> None:
    n_ch, n_freq, n_frames = 3, 11, 20
    W = np.ones((n_freq, n_ch), dtype=np.complex128)
    X = np.ones((n_ch, n_freq, n_frames), dtype=np.complex128)
    Y = apply_beamformer(W, X)
    assert Y.shape == (n_freq, n_frames)
    # W^H · X with all-ones gives n_ch * 1.0 per (f, t).
    np.testing.assert_allclose(Y, float(n_ch), atol=1e-12)


# --------------------------------------------------------------------- #
# Batch ``process`` end-to-end
# --------------------------------------------------------------------- #
def test_process_matrix_input_returns_matching_length() -> None:
    cfg = default()
    cfg.mwf.method = "gev"
    cfg.mwf.postfilter = True
    vad_audio, mic_mat = _synth_scene(cfg.fs, 1.0)
    y = Mwf(cfg).process(vad_audio, mic_mat)
    assert y.ndim == 1
    assert y.size <= mic_mat.shape[0]
    assert np.all(np.isfinite(y))


def test_process_list_input_truncates_to_shortest() -> None:
    cfg = default()
    cfg.mwf.method = "gev"
    vad_audio, mic_mat = _synth_scene(cfg.fs, 1.0)
    mic_list = [mic_mat[:, 0], mic_mat[:-100, 1], mic_mat[:, 2]]
    y = Mwf(cfg).process(vad_audio, mic_list)
    assert y.ndim == 1
    assert np.all(np.isfinite(y))


def test_method_dispatch_runs_all_three() -> None:
    fs = default().fs
    vad_audio, mic_signals = _synth_scene(fs, 0.8)
    for method in ("gev", "mwf", "mvdr"):
        cfg = default()
        cfg.mwf.method = method
        y = Mwf(cfg).process(vad_audio, mic_signals)
        assert np.all(np.isfinite(y)), f"{method} produced non-finite samples"
        assert float(np.max(np.abs(y))) > 0


def test_postfilter_toggle_changes_output() -> None:
    fs = default().fs
    vad_audio, mic_signals = _synth_scene(fs, 0.8)

    cfg1 = default()
    cfg1.mwf.method = "gev"
    cfg1.mwf.postfilter = True
    cfg2 = default()
    cfg2.mwf.method = "gev"
    cfg2.mwf.postfilter = False
    y1 = Mwf(cfg1).process(vad_audio, mic_signals)
    y2 = Mwf(cfg2).process(vad_audio, mic_signals)
    L = min(y1.size, y2.size)
    assert np.linalg.norm(y1[:L] - y2[:L]) > 1e-6
